src/js/fable-core/Array2.fs

```diff
@@ -8,9 +8,25 @@ type ArrayCons =
     [<Emit("$0.from($1)")>]
     abstract From: 'T seq -> 'T array
 
-let map (f: 'T->'U) (source: 'T[]) (cons: ArrayCons) =
+let mapIndexed (f: int -> 'T -> 'U) (source: 'T[]) (cons: ArrayCons) =
+    let len = source.Length
+    let target = cons.Create(len)
+    for i = 0 to (len - 1) do
+        target.[i] <- f i source.[i]
+    target
+
+let map (f: 'T -> 'U) (source: 'T[]) (cons: ArrayCons) =
+    // TODO: Should we involve currying and re-use mapIndexed here?
+    // Don't quite know yet how is it optimized so can't decide on performance vs code reuse.
+    // mapIndexed (fun _ x -> f x) source
```

Comment:
You can run `sh build.sh FableCoreJS` to see the generated code, [or just use the REPL](http://fable.io/repl/#?code=open%20Fable.Core%0A%0Atype%20ArrayCons%20%3D%0A%20%20%20%20%5B%3CEmit(%22new%20%240(%241)%22)%3E%5D%0A%20%20%20%20abstract%20Create%3A%20capacity%3A%20int%20-%3E%20'T%20array%0A%20%20%20%20%5B%3CEmit(%22%240.from(%241)%22)%3E%5D%0A%20%20%20%20abstract%20From%3A%20'T%20seq%20-%3E%20'T%20array%0A%0Alet%20mapIndexed%20(f%3A%20int%20-%3E%20'T%20-%3E%20'U)%20(source%3A%20'T%5B%5D)%20(cons%3A%20ArrayCons)%20%3D%0A%20%20%20%20let%20len%20%3D%20source.Length%0A%20%20%20%20let%20target%20%3D%20cons.Create(len)%0A%20%20%20%20for%20i%20%3D%200%20to%20(len%20-%201)%20do%0A%20%20%20%20%20%20%20%20target.%5Bi%5D%20%3C-%20f%20i%20source.%5Bi%5D%0A%20%20%20%20target%0A%0Alet%20map%20(f%3A%20'T%20-%3E%20'U)%20(source%3A%20'T%5B%5D)%20(cons%3A%20ArrayCons)%20%3D%0A%20%20%20%20mapIndexed%20(fun%20i%20x%20-%3E%20f%20x)%20source%20cons%0A&html=%3Chtml%3E%3Cbody%3E%3C%2Fbody%3E%3C%2Fhtml%3E%0A). As you can see, in this case it should be fine as the generated is very simple. But we need to be careful with curried function in fable-core as they can get slow if the compiler wraps them with `CurriedLambda`.

Reply:
Thanks for the tip about the REPL (impressive stuff, btw). Yeah, I saw `CurriedLambda` in the output, but thought of possibility of it being optimized-away in the babel/rollup/webpack/whatever minification pipeline. I'll avoid it in this particular casein (and minimize usage in fable-core).